go/vt/topo/zk2topo/lock.go

```diff
@@ -39,7 +43,45 @@ type zkLockDescriptor struct {
 
 // Lock is part of the topo.Conn interface.
 func (zs *Server) Lock(ctx context.Context, dirPath, contents string) (topo.LockDescriptor, error) {
-	// Lock paths end in a trailing slash to that when we create
+	return zs.lock(ctx, dirPath, contents)
+}
+
+// TryLock is part of the topo.Conn interface.
+// TryLock provides exactly same functionality as 'Lock', the only difference is
+// it tires its best to be unblocking call. Unblocking is the best effort though.
+// If there is already lock exists for dirPath then TryLock
+// unlike Lock will return immediately with error 'lock already exists'.
+func (zs *Server) TryLock(ctx context.Context, dirPath, contents string) (topo.LockDescriptor, error) {
+	// We list all the entries under dirPath
+	entries, err := zs.List(ctx, dirPath)
+	if err != nil {
+		// We need to return the right error codes, like
+		// topo.ErrNoNode and topo.ErrInterrupted, and the
+		// easiest way to do this is to return convertError(err).
+		// It may lose some of the context, if this is an issue,
+		// maybe logging the error would work here.
+		return nil, convertError(err, dirPath)
+	}
+
+	// if there is a folder '/locks' with some entries in it then we can assume that keyspace already have a lock
+	// throw error in this case
+	for _, e := range entries {
+		path := string(e.Key[:])
+		if nodeUnderLockPath.MatchString(path) {
+			return nil, topo.NewError(topo.NodeExists, fmt.Sprintf("lock already exists at path %s", dirPath))
+		}
+
+		// TODO: instead of list should I call listDir and assume /locks directory only exists if there is a lock
+		// TODO: Should we check if all the children under lock is ephemeral
```

Comment:
Same locks being a file/directory

Reply:
apart from consul every other implementation looks for folder. So we should be good here